### ISSUE TITLE
Structural termination checking is unsound

### ISSUE THREAD

**Issue by rosekunkel** [ucsd-progsys/liquidhaskell#1404]
The following passes the strucutral termination checker and is deemed safe:
```haskell
module Absurd where

{-@ absurd :: {False} @-}
absurd :: a
absurd =
  let loop x = loop x in
    loop ()

{-@ oneIsTwo :: {1 == 2} @-}
oneIsTwo = absurd
```
Running with `--no-structural` correctly returns "unsafe". Unfortunately, reverting the recent changes I made to the checker does not fix the problem, the old checker still deemed this code safe.

**Comment by rosekunkel** [comment]
More generally, functions defined in recursive let-bindings are not checked for termination.

**Comment by rosekunkel** [comment]
More unsoundness:
```haskell
module Absurd where

{-@ foo :: {False} @-}
foo = bar

{-@ bar :: {False} @-}
bar = foo

{-@ oneIsTwo :: {1 == 2} @-}
oneIsTwo = foo
```

**Comment by rosekunkel** [comment]
Interestingly, these are actually examples of the same bug. For some reason GHC compiles the above to a recursive let-binding, something like
```haskell
fooBar =
  let baz = baz in
    (baz, baz)

foo = fst fooBar
bar = snd fooBar
```
Adding a parameter to each function causes it to compile them to mutually-recursive top-level bindings, which are correctly rejected:
```haskell
module Absurd where

{-@ foo :: _ -> {False} @-}
foo :: () -> a
foo x = bar x

{-@ bar :: _ -> {False} @-}
bar :: () -> a
bar x = foo x

{-@ oneIsTwo :: {1 == 2} @-}
oneIsTwo = foo ()
```

**Comment by ranjitjhala** [comment]
Whoa. Good catch! Do you mean non toplevel Rec binders are not checked (ie
only top level RecBinders are checked?)

**Comment by rosekunkel** [comment]
Yes, essentially.

**Comment by nikivazou** [comment]
Interesting! @rosekunkel are you trying to fix it? (If not maybe we should switch to semantic termination by default, thought I do prefer the automation of structural :) )

**Comment by rosekunkel** [comment]
@nikivazou I am, but it's a pretty busy week for me so I might not have a fix up for a little while. Conceptually the fix isn't hard but we'll need to re-architect the checker a little bit.

**Comment by rosekunkel** [comment]
More unsoundness:
```haskell
{-@ LIQUID "--exactdc" @-}
module Absurd where

data Peano = Z | S Peano

{-@ impossible :: {msg:_ | False} -> a @-}
impossible :: String -> a
impossible msg = error msg

{-@ foo :: {n : Peano | n /= Z} -> _ -> {False} @-}
foo :: Peano -> Peano -> a
foo (S n) m = bar n (S m)
foo _ _ = impossible "unreachable"

{-@ bar :: _ -> {n : Peano | n /= Z} -> {False} @-}
bar :: Peano -> Peano -> a
bar m (S n) = foo (S m) n
bar _ _ = impossible "unreachable"

{-@ oneIsTwo :: {1 == 2} @-}
oneIsTwo = foo (S Z) Z
```

This time the unsoundness comes from the handling of mutually-recursive definitions, not the fact that `let` expressions aren't checked. The logic for mutually-recursive definitions was already kind of broken, but I think it probably could not be turned into unsoundness before my changes.

**Comment by rosekunkel** [comment]
I wonder whether going forward if we should have the structural checker emit termination metrics for functions, instead of just blindly trusting its output.

It seems like that could reduce the amount of code we have to trust.

**Comment by nikivazou** [comment]
Or, we can reject all mutual recursive function as non-terminating and require semantic metrics?

**Comment by rosekunkel** [comment]
That's a good stopgap measure at least.

**Comment by fgaz** [comment]
So did #1471 fix it?

**Comment by ranjitjhala** [comment]
Yes, I *believe* that @rosekunkel 's fixes in #1471 address this unsoundness

1.  by disabling the structural checker, and
2. adding the above as neg-tests that make sure LH rejects the above programs.

Can you confirm Rose? And if so, shall we close this issue?

**Comment by rosekunkel** [comment]
Yes, this is fixed. The structural tester was essentially disabled for
functions with mutual recursion, and the other issues were fixed. This
issue can be closed.
